<?xml version="1.0"?>

<launch>
  <arg name="robot_name" default=""/>
  <arg name="maplab_sensor_calibration_file" default=""/>
  <arg name="maplab_imu_to_camera_time_offset_ns" default="0"/>
  <arg name="maplab_image_topic_suffix" default=""/>
  <arg name="maplab_camera_mask_file" default=""/>
  <arg name="maplab_apply_clahe" default="false"/>

  <arg name="maplab_selected_ncamera_sensor_id" default=""/>
  <arg name="maplab_imu_sensor_id" default=""/>
  <arg name="maplab_point_cloud_to_depth_map_conversion_camera_id" default=""/>
  <arg name="maplab_point_cloud_to_depth_map_conversion_include_intensity_image" default="true"/>

  <arg name="maplab_rovioli_localization_map_folder" default=""/>
  <arg name="maplab_maplab_node_localization_map_folder" default=""/>

  <arg name="maplab_world_frame" default="map"/>

  <arg name="maplab_verbosity" default="0"/>

  <arg name="maplab_save_map_every_n_sec" default="0"/>
  <arg name="maplab_keyframe_when_saving" default="false"/>
  <arg name="maplab_remove_bad_landmarks" default="false"/>

  <arg name="maplab_attach_images" default="false"/>
  <arg name="maplab_attach_point_clouds" default="false"/>
  <arg name="maplab_attach_point_cloud_maps" default="false"/>

  <arg name="maplab_map_folder" default="/tmp/maplab/map_$(arg robot_name)"/>

  <!-- ===================================== MAPLAB NODE =========================================== -->

  <group ns="$(arg robot_name)/maplab_node">
    <remap from="/$(arg robot_name)/odometry/maplab_odom_T_M_I" to="/$(arg robot_name)/rovioli/maplab_odom_T_M_I" unless="$(eval robot_name == '')"/>
    <remap from="/odometry/maplab_odom_T_M_I" to="/rovioli/maplab_odom_T_M_I" if="$(eval robot_name == '')"/>

		<!-- MAPLAB_NODE -->
		<include file="$(find maplab_node)/launch/maplab_node.launch">

      <!-- SENSORS -->
      <arg name="sensor_calibration_file" default="$(arg maplab_sensor_calibration_file)"/>
      <!-- only needed if there are multiple ncameras/lidars/imus/odometries in the calibration file. -->
      <arg name="selected_ncamera_sensor_id" default="$(arg maplab_selected_ncamera_sensor_id)"/>
      <arg name="selected_imu_sensor_id" default="$(arg maplab_imu_sensor_id)"/>
      <arg name="selected_lidar_sensor_id" default=""/>
      <arg name="selected_odometry_6dof_sensor_id" default=""/>

      <!-- DATA SOURCE -->
      <arg name="datasource_type" default="rostopic"/>
      <arg name="datasource_rosbag" default=""/>
      <arg name="imu_to_camera_time_offset_ns" default="$(arg maplab_imu_to_camera_time_offset_ns)"/>

      <!-- DATA PREPROCESSING -->
      <arg name="image_apply_clahe_histogram_equalization" default="$(arg maplab_apply_clahe)"/>

      <!-- LOCALIZATION BASED ON VISION -->
      <arg name="enable_visual_localization" default="true" if="$(eval maplab_maplab_node_localization_map_folder != '')"/>
      <arg name="visual_localization_map_folder" default="$(arg maplab_maplab_node_localization_map_folder)"/>

      <!-- LOCALIZATION BASED ON LIDAR -->
      <arg name="enable_lidar_localization" default="false"/>
      <arg name="lidar_localization_map_folder" default=""/>

      <!-- ONLINE MAPPING -->
      <arg name="enable_online_mapping" default="false"/>

      <!-- MAP BUILDING -->
      <arg name="map_output_folder" default="$(arg maplab_map_folder)"/>
      <arg name="map_save_on_shutdown" default="true"/>
  		<arg name="map_overwrite_enabled" default="false"/>
  		<arg name="map_compute_localization_map" default="false"/>
  		<arg name="map_initialize_and_triangulate_landmarks_when_saving" default="true"/>
  		<arg name="map_run_keyframing_when_saving" default="$(arg maplab_keyframe_when_saving)"/>
      <arg name="map_remove_bad_landmarks" default="$(arg maplab_remove_bad_landmarks)"/>
  		<arg name="map_save_every_n_sec" default="$(arg maplab_save_map_every_n_sec)"/>
  		<arg name="map_split_map_into_submaps_when_saving_periodically" default="true"/>
      <arg name="map_add_odometry_edges_if_less_than_n_common_landmarks" default="20"/>

  		<!-- ATTACHING RESOURCES TO MAP -->
  		<arg name="map_builder_save_image_as_resources" default="$(arg maplab_attach_images)"/>
  		<arg name="map_builder_save_point_clouds_as_resources" default="$(arg maplab_attach_point_clouds)"/>
      <arg name="map_builder_save_point_clouds_as_range_image_camera_id" default="$(arg maplab_point_cloud_to_depth_map_conversion_camera_id)"/>
      <arg name="map_builder_save_point_clouds_as_range_image_including_intensity_image" default="$(arg maplab_point_cloud_to_depth_map_conversion_include_intensity_image)"/>
  		<arg name="map_builder_save_point_cloud_maps_as_resources" default="$(arg maplab_attach_point_cloud_maps)"/>

  		<!-- TF -->
      <arg name="tf_map_frame" default="$(arg maplab_world_frame)"/>
      <arg name="tf_mission_frame" default="maplab_$(arg robot_name)_odom"/>
      <arg name="tf_imu_frame" default="maplab_$(arg robot_name)_imu"/>
  		<arg name="tf_imu_refined_frame" default="maplab_$(arg robot_name)_imu_refined"/>

  		<!-- DEBUGGING AND LOGGING -->
  		<arg name="alsologtostderr" default="true"/>
  		<arg name="colorlogtostderr" default="true"/>
  		<arg name="v" default="$(arg maplab_verbosity)"/>
  		<arg name="publish_debug_markers" default="true"/>
      <arg name="vis_default_namespace" default="maplab_$(arg robot_name)_rviz"/>

  		<arg name="detection_visualize_keypoints" default="true"/>
  		<arg name="feature_tracker_visualize_feature_tracks" default="true"/>
  		<arg name="feature_tracker_visualize_keypoint_matches" default="true"/>
  		<arg name="feature_tracker_visualize_keypoints" default="true"/>
  		<arg name="feature_tracker_visualize_keypoints_individual_frames" default="true"/>

      <arg name="vio_camera_topic_suffix" default="$(arg maplab_image_topic_suffix)" />
    </include>
	</group>

	<include file="$(find maplab_node)/launch/maplab_node_w_rovioli/rovioli.launch">
    <arg name="rovioli_sensor_calibration_file" default="$(arg maplab_sensor_calibration_file)"/>
    <arg name="robot_name" default="$(arg robot_name)"/>
    <arg name="rovioli_imu_to_camera_time_offset_ns" default="$(arg maplab_imu_to_camera_time_offset_ns)"/>
    <arg name="rovioli_apply_clahe" default="false"/>
    <arg name="rovioli_world_frame" default="darpa"/>
    <arg name="rovioli_verbosity" default="$(arg maplab_verbosity)"/>
    <arg name="rovioli_image_topic_suffix" default="$(arg maplab_image_topic_suffix)"/>
    <arg name="rovioli_visualize" default="true"/>
    <arg name="rovioli_enable_health_checking" default="false"/>
  </include>
</launch>
